### PR TITLE
Sh/model

### PR DESCRIPTION
**Pull request by sangHa0411**
1. 수정방향
    1. Tagging을 고민 하였을 때 어차피 처음만 B이고 나머지가 I인 것을 확인하였습니다.
    2. 또한 B-PS와 I-PS는 연관된 의미를 가지고 있을 탠데 아예 다른 클래스로 구분하는 것이 조금은 적절하지 않다고 판단했습니다.
    3. 따라서 I-PS와 같은 태그는 후처리로 따로 처리하도록 하고 I-PS와 같은 태그들을 B-PS로 변경함으로써 전처리를 진행하였습니다.

2. 변경방향
    1. 13개의 Label을 7개의 Label으로 줄여서 token classification을 진행
    2. 학습을 7개의 label을 분류하는 식으로 진행하고 evaluation을 할 때 후처리를 함으로써 I-PS, I-DT와 같은 태그를 만들어주면서 후처리, Metric을 계산
 
3. 모델
    1. 성능 감소 
        1. LSTM Head
        2. Vstack (Output의 Hidden States들을 concat 혹은 add) 하는 구조
    2. 성능 향상
        1. Linear head를 2개를 만들고 해당 head를 통해서 구한 2개의 vector를 concat해서 최종 classifiction layer로 전달하는 구조
        2. 저희 MRC대회 했을 때 얘기나온 multiple dropout 구조를 약간 변형해서 만들었습니다.
        
            image

        ```python
        class RobertaWithLinearHead(RobertaPreTrainedModel):
            _keys_to_ignore_on_load_unexpected = [r"pooler"]
            _keys_to_ignore_on_load_missing = [r"position_ids"]
        
            def __init__(self, config):
                super().__init__(config)
                self.num_labels = config.num_labels
        
                self.roberta = RobertaModel(config, add_pooling_layer=False)
                classifier_dropout = (
                    config.classifier_dropout if config.classifier_dropout is not None else config.hidden_dropout_prob
                )
                self.dropout = nn.Dropout(classifier_dropout)
                self.classifier1 = nn.Linear(config.hidden_size, config.hidden_size)
                self.classifier2 = nn.Linear(config.hidden_size, config.hidden_size)
        
                self.classifier = nn.Linear(config.hidden_size*2, config.num_labels)
                self.post_init()
        
            def forward(
                self,
                input_ids: Optional[torch.LongTensor] = None,
                attention_mask: Optional[torch.FloatTensor] = None,
                token_type_ids: Optional[torch.LongTensor] = None,
                position_ids: Optional[torch.LongTensor] = None,
                head_mask: Optional[torch.FloatTensor] = None,
                inputs_embeds: Optional[torch.FloatTensor] = None,
                labels: Optional[torch.LongTensor] = None,
                output_attentions: Optional[bool] = None,
                output_hidden_states: Optional[bool] = None,
                return_dict: Optional[bool] = None,
            ) -> Union[Tuple[torch.Tensor], TokenClassifierOutput]:
        
                return_dict = return_dict if return_dict is not None else self.config.use_return_dict
        
                outputs = self.roberta(
                    input_ids,
                    attention_mask=attention_mask,
                    token_type_ids=token_type_ids,
                    position_ids=position_ids,
                    head_mask=head_mask,
                    inputs_embeds=inputs_embeds,
                    output_attentions=output_attentions,
                    output_hidden_states=output_hidden_states,
                    return_dict=return_dict,
                )
        
                sequence_output = outputs[0]
                sequence_output = self.dropout(sequence_output)
        
                logits1 = self.classifier1(sequence_output)
                logits2 = self.classifier2(sequence_output)
        
                logits = torch.cat([logits1, logits2], dim=-1)
                logits = self.classifier(logits)
        
                loss = None
                if labels is not None:
                    loss_fct = CrossEntropyLoss()
                    loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
        
                if not return_dict:
                    output = (logits,) + outputs[2:]
                    return ((loss,) + output) if loss is not None else output
        
                return TokenClassifierOutput(
                    loss=loss,
                    logits=logits,
                    hidden_states=outputs.hidden_states,
                    attentions=outputs.attentions,
                )

        ```

최종 성능
1. PLM : klue/roberta-large
2. wandb : f1 score
image

### CODE REVIEW

**Comment by snoop2head** [comment]
수고하셨습니다!!! README도 해당 업데이트 반영하도록 하겠습니다 ㅎㅎ
